doc/ccAPPSctl.1
.\"
.\" Copyright (C) 2009-2013 by ccPAAS bv
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in all
.\" copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.\" SOFTWARE.
.\"
.TH "ccPAAS" "1" "April 2023" "ccPAAS" ""
.SH "NAME"
ccPAASctl \- open source production planning and scheduling
.br
.SH "SYNOPSIS"
.B ccPAASctl subcommand [options] [arguments]

.SH "DESCRIPTION"
ccPAAS is an open source production planning and scheduling application.
.P
This program is a wrapper for the administrative actions of the
application.

.SH OPTIONS
.TP
.BI "subcommand"
Name of the command to execute. Use 'ccPAASctl help' to get a list of
available subcommands.
.TP
.BI "[options] [arguments]"
Options and arguments applied to the chosen subcommand. Use 'ccPAASctl
help subcommand' to get a list of options and arguments applicable to a
certain subcommand.

.SH ENVIRONMENT
.TP
ccPAAS_HOME
Optionally points to a directory where the initialization
files will be searched.

.SH RETURNS
0 when successful, non-zero in case of errors

.SH "AUTHORS"
.B ccPAAS
is developed as Open Source under the MIT license.

The authors and contributors are listed on http://ccPAAS.com.


doc/release-notes.rst
Release notes
-------------

9.16.0 (Upcoming release)
========================

This release is scheduled for mid May 2026.
You can already check out a `preview <https://demo-preview.ccAPPS.com>`_.

9.15.0 (2026-04-10)
===================

.. rubric:: User interface

- | Supply path refactoring: The supply path has been redesigned to provide a
    clearer visualization of the supply chain. Levels can be collapsed and expanded,
    objects are draggable...

.. rubric:: Production planning

- | Manufacturing orders now respect the batch window of the operations.
  | When requirements are close to each other, the algorithm now
    plans a single producing manufacturing order to cover the requirements.
  | In this release this feature is only active when running a plan without any
    constraints. In future releases we will also support this when running a plan
    with constraints.

- | Bug fix: Some models experienced issues with excess proposed supply in the
    presence of operation multiples and temporary unresolvable material shortages.

- | Bug fix: The pegging between supply and demand would be partially missing when
    the buffer has periods with negative inventory.

- | Extra robustness to deal with negative safety stocks.
  | Some people throw ugly data to ccAPPS. We treat negatives safety stocks values now
    as a zero.

.. rubric:: Odoo integration

- | 18, 19: The default behavior for bringing draft purhase orders (aka RFQs) into
    ccAPPS has been adjusted.
  | Earlier releases didn't bring them in.
  | Now we bring them in to reduce the risk of re-proposing a new, duplicate RFQ.

9.14.0 (2026-03-06)
===================

.. rubric:: User interface

- | Emails sent upon success or failure of a scheduled workflow will now contain
    a table summarizing all the tasks executed during the workflow plus some useful
    links. This allows a better overview of the workflow status from that email.

.. rubric:: Odoo integration

- | We have added branches 17.0-dev, 18.0-dev and 19.0-dev
    in the connector github repository.
  | This allows us to separate active development from the stable version
    of the connector.

- | 18, 19: Changed the behavior when a bill of material doesn't specify
    an operation on which a component needs to be consumed.
  | Odoo then consumes the component on the LAST operation.
  | Earlier version of the connector put the consumption of the component
    on the FIRST operation.
  | This difference created more confusion and issues than resolving
    issues. The connector now follows the odoo logic.

- | 18, 19: The calculation of the reserved stocks has been adjusted to
    account for more complex odoo configurations with stock pickings.

- | 17, 18, 19: Bug fix for the bulk export from ccAPPS to odoo in multi-company
    setups.

- | 18, 19: Improved mapping for subcontractor suppliers to account for date
    effectivity and price breaks.

- | 17, 18, 19: For readability reasons, the *Vendor*, *Manufacturing Order* and
    *sales order line* columns of the Recommendation table were merged into a
    single column called "Related data".

- | 17, 18, 19: For security reasons, a new field called "ccAPPS interface user" was added
    to the ccAPPS settings in Odoo. Only that user can run an "import from Odoo" task
    in ccAPPS.

- | 18, 19: Refined logic to get the due date of a sales order line.
  | We now use in order of preference the date of the delivery stock move, the committed
    date on the sales order line, the committed date on the sale order, or the ordering
    date.

- | 18, 19: Support for production by lot or unique serial number.
  | Manufacturing orders for such products need a specific calculation of the
    intended production quantity.

- | 17, 18, 19: Bug fix when exporting POs/MOs/DOs from the sales order table. The exported
    quantity could be incorrect if some rows were unselected.

.. rubric:: Production planning

- | Performance improvements on models with complex calendars (typically for bucketized
    capacity) and when running in UTC timezone.

.. rubric:: System administration

- | Extra robustness was added to assure the system isn't waiting forever on
    looping or blocked tasks. We now cancel and clean up unhealty tasks.

.. rubric:: Upgrade compatibility heads-up

| This ia a new section in the release notes!
| It provides a heads-up to specific changes that can impact your customizations or
  integrations when you upgrade to this new release.

- | Applicable if you customized the kanban card template "input/kanban_card.html":
  | You need to create a new template "input/kanbancard.html", this time using vue
    syntax rather than the angularjs syntax.

9.13.0 (2026-01-31)
===================

.. rubric:: Production planning

- | The number of problems generated by the planning algorithm has been
    drastically reduced.
  | Experience learns that these alerts aren't very efficient in helping
    planners to review or resolve issues in the plan.
  | ccAPPS now no longer generates problems for: unplanned demands,
    demands planned late, demands planned early, operations before current,
    operations before the fence.
  | The user interface provides enough alternatives to identify and review
    these.

- | The logic to create the "why short or late" information has been
    refactored to be easier to understand.

.. rubric:: Odoo integration

- | Added support for Odoo 19.
  | The connector is still experimental and for testing purposes only.
  | With Odoo 19 the connector has been refactored into a real next-generation.

- | It's now possible to run the Odoo import as part of the plan generation.
  | In practice, you'll always run the plan right after the import, isn't it?
    So, we made it now a checkbox in the plan generation options.

- | A link back to Odoo is now displayed in the upper right corner.

- | When exporting manufacturing orders from ccAPPS to Odoo, the connectors are

ccAPPSctl.py
#!/usr/bin/env python3

#
# Copyright (C) 2007-2013 by ccAPPS bv
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

r"""
This command is the wrapper for all administrative actions on ccAPPS.
"""

import os
import sys
import site

if __name__ == "__main__":
    try:
        # Autodetect Python virtual enviroment
        venv = os.environ.get("VIRTUAL_ENV", None)
        if not venv:
            curdir = os.path.dirname(os.path.realpath(__file__))
            for candidate in (
                # Development layout
                os.path.join(curdir, "venv"),
                # Linux install layout
                os.path.join(curdir, "..", "share", "ccAPPS", "venv"),
            ):
                if os.path.isfile(
                    os.path.join(candidate, "bin", "python3")
                ) and os.path.isfile(os.path.join(candidate, "bin", "activate")):
                    os.environ["VIRTUAL_ENV"] = candidate
                    venv = candidate
                    break

        # Activate Python virtual environment
        if venv:
            prev_length = len(sys.path)
            os.environ["PATH"] = os.pathsep.join(
                [os.path.join(venv, "bin")]
                + os.environ.get("PATH", "").split(os.pathsep)
            )
            path = os.path.realpath(
                os.path.join(
                    venv,
                    "lib",
                    "python%d.%d" % sys.version_info[:2],
                    "site-packages",
                )
            )
            site.addsitedir(path)
            sys.path[:] = sys.path[prev_length:] + sys.path[0:prev_length]
            sys.real_prefix = sys.prefix
            sys.prefix = venv

        # Initialize django
        os.environ.setdefault("DJANGO_SETTINGS_MODULE", "ccAPPSdb.settings")
        import django

        django.setup()

        # Synchronize the scenario table with the settings
        from ccAPPSdb.common.models import Scenario

        Scenario.syncWithSettings()

        if "version" in sys.argv or "--version" in sys.argv:
            # Display the version and stop
            import ccAPPSdb

            sys.stdout.write(ccAPPSdb.__version__ + "\n")
        else:
            # Run the command
            from django.core.management import execute_from_command_line

            execute_from_command_line(sys.argv)

    except KeyboardInterrupt:
        print("\nInterrupted with Ctrl-C")
        sys.exit(1)
